### PR TITLE
Inventory unit tests

### PR DESCRIPTION
**Pull request by jcanocan**
**What this PR does / why we need it**:
It refactors several fixtures to allow the use of any inventory mock. The affected fixtures are:
- host_vars
- add_group
- add_child
- add_host

It also adds unit tests for the following inventory functions:
- set_composable_vars
- format_dynamic_api_exc

Finally, it increases test coverage for `get_vmis_for_namespace` function.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # [CNV-41852](https://issues.redhat.com/browse/CNV-41852)

**Special notes for your reviewer**:

Follow-up of PR #96 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 # GNU General Public License v3.0+ (see LICENSE or https://www.gnu.org/licenses/gpl-3.0.txt)
 
 from __future__ import absolute_import, division, print_function
+import json
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Done.

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 2]
```diff
@@ -15,6 +16,7 @@
     GetVmiOptions,
     InventoryModule,
     KubeVirtInventoryException,
+    DynamicApiError
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Done.

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 3]
```diff
@@ -223,44 +241,47 @@ def inventory(mocker):
...
+def host_vars(monkeypatch, request):
     host_vars = {}
 
     def set_variable(host, key, value):
         if host not in host_vars:
             host_vars[host] = {}
         host_vars[host][key] = value
 
-    monkeypatch.setattr(inventory.inventory, "set_variable", set_variable)
+    monkeypatch.setattr(request.getfixturevalue(request.param).inventory, "set_variable", set_variable)
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Dropped this approach in favor of a separate file for the tests that require another mock inventory object.

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 4]
```diff
@@ -402,15 +423,15 @@ def test_is_windows(inventory, guest_os_info, annotations, expected):
...
-        ({"vmis": [WINDOWS_VMI_3]}, WINDOWS_VMI_3, True),
-        ({"vmis": [WINDOWS_VMI_4]}, WINDOWS_VMI_4, True),
+        ({"vmis": [BASIC_VMI]}, BASIC_VMI, "inventory", False),
+        ({"vmis": [WINDOWS_VMI_1]}, WINDOWS_VMI_1, "inventory", True),
+        ({"vmis": [WINDOWS_VMI_2]}, WINDOWS_VMI_2, "inventory", True),
+        ({"vmis": [WINDOWS_VMI_3]}, WINDOWS_VMI_3, "inventory", True),
+        ({"vmis": [WINDOWS_VMI_4]}, WINDOWS_VMI_4, "inventory", True),
     ],
-    indirect=["client"],
+    indirect=["client", "host_vars"],
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Dropped.

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 5]
```diff
@@ -879,17 +921,39 @@ def windows_vmi_host_vars(basic_vmi_host_vars):
...
+            "children_group_empty",
+            False,
+            None,
+            "base_vmi_host_vars",
+            False,
+            False,
+            "inventory",
+            "inventory",
+            "inventory",
+            "inventory",
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Droped.

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 6]
```diff
@@ -1141,3 +1257,163 @@ def test_set_ansible_host_and_port(monkeypatch, request, inventory, host_vars, v
...
+def vmi_host_vars():
+    return {
+        "vmi_migration_method": "BlockMigration",
+        "vmi_guest_os_info": {"id": "fedora", "version": "39"},
+        "cluster_name": {"test-cluster"},
+    }
+
+
+@pytest.fixture(scope="function")
+def inventory_composable_vars(mocker):
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Dropped.

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 7]
```diff
@@ -1141,3 +1257,163 @@ def test_set_ansible_host_and_port(monkeypatch, request, inventory, host_vars, v
...
+        "testvmi",
+        "fedora_39",
+    ]
+
+
+@pytest.mark.parametrize(
+    "host_vars,add_group,add_child,add_host",
+    [
+        (
+            "inventory_composable_vars",
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Dropped.

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 8]
```diff
@@ -269,7 +290,7 @@ def add_child(group, name):
         if name not in children[group]:
             children[group].append(name)
 
-    monkeypatch.setattr(inventory.inventory, "add_child", add_child)
+    monkeypatch.setattr(request.getfixturevalue(request.param).inventory, "add_child", add_child)
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Dropped.

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 9]
```diff
@@ -223,44 +241,47 @@ def inventory(mocker):
...
+def add_group(monkeypatch, request):
     groups = []
 
     def add_group(name):
         if name not in groups:
             groups.append(name)
+        return name
 
-    monkeypatch.setattr(inventory.inventory, "add_group", add_group)
+    monkeypatch.setattr(request.getfixturevalue(request.param).inventory, "add_group", add_group)
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Dropped.

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 10]
```diff
@@ -6,12 +6,14 @@
 
 __metaclass__ = type
 
+import copy
+import json
```
Comment (0xFelix) [inline tests/unit/plugins/inventory/test_kubevirt.py:10]: Can you import just the needed functions from those packages? (`from ... import ...`)

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 11]
```diff
@@ -1143,3 +1145,65 @@ def test_set_ansible_host_and_port(monkeypatch, request, inventory, host_vars, v
...
     assert request.getfixturevalue(result) == host_vars
+
+
+@pytest.fixture(scope="function")
+def body_error(mocker):
+    error = DynamicApiError(e=mocker.Mock())
+    error.headers = None
+
+    body = "This is a test error"
+    error.body = body
```
Comment (0xFelix) [inline tests/unit/plugins/inventory/test_kubevirt.py:1238]: Merge those lines?

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 12]
```diff
@@ -1143,3 +1145,65 @@ def test_set_ansible_host_and_port(monkeypatch, request, inventory, host_vars, v
...
+        "Content-Type": "application/json"
+    }
+
+    body = {
+        "message": "This is a test error"
+    }
+
+    json_data = json.dumps(body)
+    byte_array = json_data.encode('utf-8')
+    error.body = byte_array
```
Comment (0xFelix) [inline tests/unit/plugins/inventory/test_kubevirt.py]: Merge those lines?

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 13]
```diff
@@ -1143,3 +1145,65 @@ def test_set_ansible_host_and_port(monkeypatch, request, inventory, host_vars, v
...
+def status_reason_error(mocker):
+    error = DynamicApiError(e=mocker.Mock())
+    error.body = None
+    error.status = 404
+    error.reason = "This is a test error"
+    return error
+
+
+@pytest.fixture(scope="module")
+def status_reason_error_return():
```
Comment (0xFelix) [inline tests/unit/plugins/inventory/test_kubevirt.py:1272]: Do we really need separate fixtures for simple strings? I'd just define the in the parametrize block below.

File: tests/unit/plugins/inventory/test_kubevirt_composable_vars.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+@pytest.fixture(scope="module")
+def expected_groups():
+    return [
+        "vmi_node_groups",
+        "fedora_39",
+    ]
+
+
+@pytest.fixture(scope="module")
+def expected_children():
```
Comment (0xFelix) [inline tests/unit/plugins/inventory/test_kubevirt_composable_vars.py:115]: I don't think we need separate fixtures here either. Just define/comparte them directly in the test.

**Review by 0xFelix — COMMENTED** [review]
I think you should reduce the amount of fixtures where possible. It is okay to define expected results simply in a test (if there is no parametrize), or depending on the size of the expected result it is also totally okay to define them in a parametrize block.

File: tests/unit/plugins/inventory/test_kubevirt.py (hunk 10 above)
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Done.

File: tests/unit/plugins/inventory/test_kubevirt.py (hunk 11 above)
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py:1238]: Done.

File: tests/unit/plugins/inventory/test_kubevirt.py (hunk 12 above)
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Done.

File: tests/unit/plugins/inventory/test_kubevirt.py (hunk 13 above)
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: You are right! I've fallen into the bad habit of defining everything as fixtures. 
Fixed.

File: tests/unit/plugins/inventory/test_kubevirt_composable_vars.py (hunk 14 above)
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt_composable_vars.py]: Done.

**Review by 0xFelix — COMMENTED** [review]
Can you try and see if you can cut down them amount of fixtures a bit further? E.g. like for `connection_none`, `connection_not_list`, etc.? Also, could you please try to split tests into multiple files? An evergrowing file with more and more lines eventually becomes impossible to maintain.

**Comment by jcanocan** [comment]
> Can you try and see if you can cut down them amount of fixtures a bit further? E.g. like for `connection_none`, `connection_not_list`, etc.? Also, could you please try to split tests into multiple files? An evergrowing file with more and more lines eventually becomes impossible to maintain.

Same as: https://github.com/kubevirt/kubevirt.core/pull/107#discussion_r1633454201

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 15]
```diff
@@ -9,9 +9,11 @@
 import pytest
 
 from addict import Dict
-import copy
+from copy import deepcopy
```
[the remark this answers is not in the document]
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py:10]: Done.

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 16]
```diff
@@ -6,12 +6,14 @@
 
 __metaclass__ = type
 
-import pytest
-
 from addict import Dict
```
Comment (0xFelix) [inline tests/unit/plugins/inventory/test_kubevirt.py:9]: `addict` is not in the stdlib. :)

**Comment by jcanocan** [comment]
Hi @0xFelix!
I've dropped as many fixtures as possible from the codebase.
PTAL!

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 17]
```diff
@@ -125,35 +118,31 @@
...
+    BASE_SERVICE, {"spec": {"type": "LoadBalancer"}}
+)
 BASE_LOADBALANCER_SERVICE = merge_dicts(
-    BASE_SERVICE,
-    {
-        "spec": {
-            "type": "LoadBalancer"
-        }
-    }
+    BASE_SERVICE, {"spec": {"type": "LoadBalancer"}}
```
Comment (0xFelix) [inline tests/unit/plugins/inventory/test_kubevirt.py:145]: Those are the same?

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 18]
```diff
@@ -183,28 +172,115 @@
...
-        }
-    }
+    },
 )
+LOADBALANCER_SERVICE = {
+    "apiVersion": "v1",
+    "kind": "Service",
+    "metadata": {"name": "test-service"},
+    "spec": {
+        "selector": {"kubevirt.io/domain": "test-domain"},
```
Comment (0xFelix) [inline tests/unit/plugins/inventory/test_kubevirt.py:182]: Similar to the one above + selector?

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 19]
```diff
@@ -422,716 +501,561 @@ def test_ansible_connection_winrm(inventory, host_vars, client, vmi, expected):
...
+            "test-hostname",
+        ),
+        ({"spec": {"type": "NodePort"}}, "test-nodename", "test-nodename"),
     ],
 )
-def test_get_host_from_service(request, inventory, service, node_name, expected):
-    result = inventory.get_host_from_service(request.getfixturevalue(service), node_name)
+def test_get_host_from_service(inventory, service, node_name, expected):
+    result = inventory.get_host_from_service(service, node_name)
     assert result == expected
```
Comment (0xFelix) [inline tests/unit/plugins/inventory/test_kubevirt.py:541]: Could be merged

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 20]
```diff
@@ -422,716 +501,561 @@ def test_ansible_connection_winrm(inventory, host_vars, client, vmi, expected):
...
-                         base_domain="test-domain",
-                         network_name="test-interface"
-                         )
+    return GetVmiOptions(
+        use_service=True,
+        create_groups=True,
+        append_base_domain=True,
+        base_domain="test-domain",
+        network_name="test-interface",
+    )
```
Comment (0xFelix) [inline tests/unit/plugins/inventory/test_kubevirt.py:609]: Inline that into parametrize below?

File: tests/unit/plugins/inventory/test_kubevirt.py [hunk 21]
```diff
@@ -422,716 +501,561 @@ def test_ansible_connection_winrm(inventory, host_vars, client, vmi, expected):
...
+    add_child,
+    client,
+    groups,
+    vmi_group,
+    child_group,
+    create_groups,
+    network_name,
+    expected_host_vars,
+    call_functions,
+    windows,
```
Comment (0xFelix) [inline tests/unit/plugins/inventory/test_kubevirt.py:903]: Still torn about this, it just doesn't look right to me. Is there nothing we can do about it?

File: tests/unit/plugins/inventory/test_kubevirt_composable_vars.py [hunk 22]
```diff
@@ -0,0 +1,119 @@
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2021, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
Comment (0xFelix) [inline tests/unit/plugins/inventory/test_kubevirt_composable_vars.py:3]: Wrong license

**Review by 0xFelix — COMMENTED** [review]
We're getting in the right direction, but I still see the potential to split the tests up into more files.

File: tests/unit/plugins/inventory/test_kubevirt.py (hunk 17 above)
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Nice catch. I've messed up a bit with the git rebase :sweat:

File: tests/unit/plugins/inventory/test_kubevirt.py (hunk 18 above)
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: You right. Fixed.

File: tests/unit/plugins/inventory/test_kubevirt.py (hunk 19 above)
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Done.

File: tests/unit/plugins/inventory/test_kubevirt.py (hunk 20 above)
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py]: Done.

File: tests/unit/plugins/inventory/test_kubevirt_composable_vars.py (hunk 22 above)
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt_composable_vars.py]: Done.

File: tests/unit/plugins/inventory/test_kubevirt.py (hunk 21 above)
Comment (jcanocan) [reply tests/unit/plugins/inventory/test_kubevirt.py:893]: To be addressed later on in a follow-up PR.

**Review by 0xFelix — COMMENTED** [review]
Let's merge this for now to unblock work on new features. But there's still a lot to improve here.

/approve
/lgtm

**Pull request merged**